numpy/core/src/multiarray/dtype_transfer.c

```diff
@@ -2611,313 +2670,74 @@ get_fields_transfer_function(int aligned,
 
         return NPY_SUCCEED;
     }
-    /* Copy the value of the first field to dst */
-    else if (!PyDataType_HASFIELDS(dst_dtype)) {
-        names = src_dtype->names;
-        names_size = PyTuple_GET_SIZE(src_dtype->names);
 
-        /*
-         * If DECREF is needed on source fields, may need
-         * to process all the fields
-         */
-        if (move_references && PyDataType_REFCHK(src_dtype)) {
-            field_count = names_size + 1;
-        }
-        else {
-            field_count = 1;
-        }
-        structsize = sizeof(_field_transfer_data) +
-                        field_count * sizeof(_single_field_transfer);
-        /* Allocate the data and populate it */
-        data = (_field_transfer_data *)PyArray_malloc(structsize);
-        if (data == NULL) {
-            PyErr_NoMemory();
-            return NPY_FAIL;
-        }
-        data->base.free = &_field_transfer_data_free;
-        data->base.clone = &_field_transfer_data_clone;
-        fields = &data->fields;
-
-        key = PyTuple_GET_ITEM(names, 0);
-        tup = PyDict_GetItem(src_dtype->fields, key);
-        if (!PyArg_ParseTuple(tup, "Oi|O", &src_fld_dtype,
-                                                &src_offset, &title)) {
-            PyArray_free(data);
-            return NPY_FAIL;
-        }
-        field_count = 0;
-        /*
-         * Special case bool type, the existence of fields implies True
```

Comment:
Perhaps. But (famous last words) I think it is such an obscure special case I don't think it will affect anyone.

(The fact that MA code depended on this behavior in one place might say otherwise, but I still feel pretty confident about that).

To clarify the discussion, this changes the fact that in numpy master
```python
>>> a = np.zeros(4, dtype='bool')
>>> b = np.zeros(4, dtype='i4,i4')
>>> a[:] = b
>>> a
b = np.zeros(4, dtype='i4,i4')
```

In this PR, this code now raises a ValueError.